File: src/stm/dedupe.py
"""把「可信重複群組」轉成刪除計畫,並批次執行刪除。

刪除計畫(``plan_deletions``)是純邏輯,可完整單元測試與 dry-run 預覽;
實際刪除(``execute_deletions``)才碰 Spotify API。
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Iterable, Protocol

from .models import Track

# added_at 缺失時排到最後(視為最新),才不會被誤選為「最早收藏」
_MAX_SENTINEL = "￿"


@dataclass(frozen=True)
class GroupResolution:
    """單一重複群組的決議:保留 keep,刪除 remove。"""

    keep: Track
    remove: list[Track]


@dataclass(frozen=True)
class DeletionPlan:
    """整份刪除計畫,涵蓋多個重複群組。"""

    resolutions: list[GroupResolution]

    @property
    def delete_ids(self) -> list[str]:
        return [t.id for r in self.resolutions for t in r.remove]

    @property
    def is_empty(self) -> bool:
        return not self.delete_ids


def _popularity_key(track: Track):
    # 取 min:人氣高者優先;同人氣時最早收藏優先;再以 id 確保確定性
    return (-track.popularity, track.added_at or _MAX_SENTINEL, track.id)


def _oldest_key(track: Track):
    return (track.added_at or _MAX_SENTINEL, track.id)


_KEEP_STRATEGIES = {
    "popularity": _popularity_key,
    "oldest": _oldest_key,
}


def plan_deletions(groups: Iterable[list[Track]], keep: str = "popularity") -> DeletionPlan:
    """為每個重複群組挑出要保留的一首,其餘列入刪除。

    keep="popularity":保留人氣最高(同分保留最早收藏)。
    keep="oldest":保留最早收藏。
    """
    try:
        key_fn = _KEEP_STRATEGIES[keep]
    except KeyError:
        raise ValueError(
            f"未知的保留策略 {keep!r},可用:{', '.join(_KEEP_STRATEGIES)}"
        ) from None

    resolutions = []
    for group in groups:
        kept = min(group, key=key_fn)
        remove = [t for t in group if t.id != kept.id]
        resolutions.append(GroupResolution(keep=kept, remove=remove))
    return DeletionPlan(resolutions=resolutions)


class SupportsRemove(Protocol):
    def remove_saved_tracks(self, ids: list[str]) -> None: ...


def _chunked(items: list[str], size: int):
    for i in range(0, len(items), size):
        yield items[i : i + size]


def execute_deletions(
    client: SupportsRemove, plan: DeletionPlan, batch_size: int = 50
) -> int:
    """依計畫批次刪除歌曲,回傳實際刪除的數量。"""
    ids = plan.delete_ids
    for chunk in _chunked(ids, batch_size):
        client.remove_saved_tracks(chunk)
    return len(ids)

File: apps/backend/src/domain/dedupe.ts
import type { KeepStrategy, Track } from "@stm/shared";

// Turn a list of "confident duplicate" groups into a deletion plan — pure logic,
// fully unit-testable and dry-runnable. The actual API calls live in the client.
// Ported from the Python `dedupe` module, with one safety enhancement: when
// choosing which copy to keep, a *playable* track always wins over an unplayable
// one, so a cleanup never strands you with the dead copy.

// When addedAt is missing, sort it LAST (treat as newest) so it is never mistaken
// for the "oldest" copy. ￿ is the largest BMP code point — bigger than any
// real ISO timestamp.
const MAX_SENTINEL = "￿";

export interface GroupResolution {
  /** The single track kept from this duplicate group. */
  keep: Track;
  /** Every other copy in the group, marked for deletion. */
  remove: Track[];
}

export interface DeletionPlan {
  resolutions: GroupResolution[];
}

/** Plain string comparison (no locale), returning -1 / 0 / 1. */
const cmpStr = (a: string, b: string): number => (a < b ? -1 : a > b ? 1 : 0);

/**
 * Order two copies so the one to KEEP sorts first (compare < 0).
 * Shared prefix for both strategies: prefer playable, then defer to the strategy.
 */
function keepComparator(strategy: KeepStrategy): (a: Track, b: Track) => number {
  return (a, b) => {
    // Never keep an unplayable copy when a playable one exists in the group.
    if (a.isPlayable !== b.isPlayable) return a.isPlayable ? -1 : 1;

    if (strategy === "popularity") {
      // Higher popularity wins; ties break toward the earliest-added copy.
      if (a.popularity !== b.popularity) return b.popularity - a.popularity;
    }
    // "oldest" (and the popularity tie-break): earliest addedAt, then id for determinism.
    const byAdded = cmpStr(a.addedAt ?? MAX_SENTINEL, b.addedAt ?? MAX_SENTINEL);
    if (byAdded !== 0) return byAdded;
    return cmpStr(a.id, b.id);
  };
}

const pickKeep = (group: Track[], cmp: (a: Track, b: Track) => number): Track =>
  group.reduce((best, t) => (cmp(t, best) < 0 ? t : best));

/**
 * For each duplicate group, keep one track and mark the rest for deletion.
 * - "popularity": keep the most popular (ties → earliest added).
 * - "oldest": keep the earliest added.
 * A playable copy is always preferred over an unplayable one, regardless of strategy.
 */
export function planDeletions(
  groups: Track[][],
  strategy: KeepStrategy = "popularity",
): DeletionPlan {
  const cmp = keepComparator(strategy);
  const resolutions = groups.map((group) => {
    const keep = pickKeep(group, cmp);
    const remove = group.filter((t) => t.id !== keep.id);
    return { keep, remove };
  });
  return { resolutions };
}

/** All track ids slated for deletion, flattened across every group. */
export const deleteIds = (plan: DeletionPlan): string[] =>
  plan.resolutions.flatMap((r) => r.remove.map((t) => t.id));

export const isEmptyPlan = (plan: DeletionPlan): boolean => deleteIds(plan).length === 0;

/** Split ids into fixed-size batches (Spotify's saved-tracks endpoints cap at 50). */
export function chunk<T>(items: T[], size: number): T[][] {
  const out: T[][] = [];
  for (let i = 0; i < items.length; i += size) out.push(items.slice(i, i + size));
  return out;
}
